File: math_mat.cc
template <class T>
class Matrix {
public:
  Matrix(int row, int col, const T* arr = NULL, const T& mod = T(0))
			: _row(row), _col(col), _mat(new T[row*col]), _mod(mod) {
		for (int i = 0; i < _row; i++)
			for (int j = 0; j < _col; j++)
				(*this)[i][j] = (arr == NULL) ? 0 : arr[i*col+j];
	}

	Matrix(const Matrix<T>& other) : _row(0), _col(0), _mat(NULL) { *this = other; }

	void operator =(const Matrix<T>& other) {
		if (_mat == NULL || _row * _col != other.row() * other.col()) {
			if (_mat != NULL) delete [] _mat;
			_mat = new T[other.row() * other.col()];
		}
		_row = other.row();
		_col = other.col();
		_mod = other.mod();
		for (int i = 0; i < _row; i++)
			for (int j = 0; j < _col; j++)
				(*this)[i][j] = other[i][j];
	}

	~Matrix() { delete [] _mat; }

	int row() const { return _row; }
	int col() const { return _col; }
	T mod() const { return _mod; }
	T* operator [](int row) const { return _mat + row * _col; }

	Matrix<T> operator *(const Matrix& other) const {
		assert(_col == other.row());
		Matrix<T> prod(_row, other.col(), NULL, _mod);
		for (int i = 0; i < _row; i++) {
			for (int j = 0; j < other.col(); j++) {
				for (int k = 0; k < _col; k++) {
					prod[i][j] += (*this)[i][k] * other[k][j];
					if ((k & 3) == 3 and _mod != 0) {
						prod[i][j] %= _mod;
					}
				}
				prod[i][j] %= _mod;
			}
		}
		return prod;
	}

	Matrix<T> unit() const {
		assert(col() == row());
		Matrix<T> mat(row(), col(), NULL, mod());
		for (int i = 0; i < row(); ++i) mat[i][i] = 1;
		return mat;
	}

	Matrix<T> pow(long long k) const {
		Matrix<T> r(unit());
		Matrix<T> a(*this);
		while (k > 0) {
			if (k & 1) r = r * a;
			if (k > 1) a = a * a;
			k /= 2;
		}
		return r;
	}

	friend ostream& operator <<(ostream& os, const Matrix<T>& mat) {
		for (int i = 0; i < mat.row(); ++i) {
			os << "| ";
			for (int j = 0; j < mat.col(); ++j) { os << mat[i][j] << ' '; }
			os << "|\n";
		}
		return os;
	}

private:
	int _row;
	int _col;
	T* _mat;
	T _mod;
};

File: math_mat.py
class Matrix:
    def __init__(self, rows): self.rows = copy.deepcopy(rows)
    def __getitem__(self, i): return self.rows[i]
    def __repr__(self): return '\n'.join(map(str, self.rows))

    def unit(self):
        n = len(self.rows)
        mat = [[0] * n for i in range(n)]
        for i in xrange(n): mat[i][i] = 1
        return Matrix(mat)

    def mul(self, other, mod=None):
        n = len(other.rows)
        m = len(other.rows[0])
        trans = zip(*other.rows)
        ret = [[0] * m for i in range(n)]
        for i, row in enumerate(ret):
            for j in xrange(m):
                a, b, t = self.rows[i], trans[j], 0
                for k in xrange(n): t += a[k] * b[k]
                row[j] = t % mod if mod else t
        return Matrix(ret)

    def pow(self, k, mod=None):
        a = Matrix(self.rows)
        r = self.unit()
        while k > 0:
            if k & 1: r = r.mul(a, mod)
            if k > 1: a = a.mul(a, mod)
            k >>= 1
        return r

    def sum(self):
        return sum(sum(r) for r in self.rows)
